File: ts/128/110.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *   this.val = (val===undefined ? 0 : val)
 *   this.left = (left===undefined ? null : left)
 *   this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

const isBalanced = function(root) {
  const height = node => {
    if (!node) return 0
    let lh = height(node.left)
    let rh = height(node.right)
    if (lh === -1 || rh === -1 || Math.abs(lh - rh) > 1) {
      return -1
    }
    return Math.max(lh, rh) + 1
  }
  return 0 <= height(root)
}

File: py3/128/110.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        def dfs(node):
            if not node: return 0
            ht_left = dfs(node.left)
            ht_right = dfs(node.right)
            if ht_left == -1 or ht_right == -1 or abs(ht_left - ht_right) > 1: return -1
            return max(ht_left, ht_right) + 1

        return dfs(root) >= 0